src/codegen/llvm.zig

```diff
@@ -4192,37 +4192,145 @@ pub const FuncGen = struct {
         if (self.liveness.isUnused(inst))
             return null;
 
+        const target = self.dg.module.getTarget();
         const ty_op = self.air.instructions.items(.data)[inst].ty_op;
         const operand = try self.resolveInst(ty_op.operand);
         const operand_ty = self.air.typeOf(ty_op.operand);
         const operand_scalar_ty = operand_ty.scalarType();
         const dest_ty = self.air.typeOfIndex(inst);
-        const dest_llvm_ty = try self.dg.llvmType(dest_ty);
+        const dest_scalar_ty = dest_ty.scalarType();
 
-        if (operand_scalar_ty.isSignedInt()) {
-            return self.builder.buildSIToFP(operand, dest_llvm_ty, "");
-        } else {
-            return self.builder.buildUIToFP(operand, dest_llvm_ty, "");
+        if (intrinsicsAllowed(dest_scalar_ty, target)) {
+            const dest_llvm_ty = try self.dg.llvmType(dest_ty);
+            if (operand_scalar_ty.isSignedInt()) {
+                return self.builder.buildSIToFP(operand, dest_llvm_ty, "");
+            } else {
+                return self.builder.buildUIToFP(operand, dest_llvm_ty, "");
+            }
         }
+
+        return self.intToFloat(operand, operand_ty, dest_ty);
+    }
+
+    fn intToFloat(
+        self: *FuncGen,
+        source: *const llvm.Value,
+        int_ty: Type,
+        float_ty: Type,
+    ) Allocator.Error!*const llvm.Value {
+        const target = self.dg.module.getTarget();
+        const int_scalar_ty = int_ty.scalarType();
+        const float_scalar_ty = float_ty.scalarType();
+
+        const opt_abi_ty = self.dg.getAtomicAbiType(int_scalar_ty, false);
+        const operand: *const llvm.Value = if (opt_abi_ty) |_| {
+            // source operand needs widening and truncating
+            @panic("TODO @intToFloat: handle integers of non-pot bitsize by widening them");
+        } else source;
+
+        const int_bits = @intCast(u16, int_scalar_ty.bitSize(target));
+        const compiler_rt_int_abbrev = compilerRtAbbrev(int_bits);
+
+        const float_bits = float_scalar_ty.floatBits(target);
+        const compiler_rt_float_abbrev = compilerRtAbbrev(float_bits);
+
+        var fn_name_buf: [64]u8 = undefined;
```

Comment:
Consider whether you want to have this be separate logic (as it is here) or whether you want to add int_to_float/float_to_int to the `buildFloatOp` function

Reply:
I was thinking of extracting common logic in `airIntToFloat` and `airFloatToInt` as we do share quite a bit there between the two, but I will look into it after I have all of your other comments addresses first, and in particular, handling of widening of input/output operands so that we have a good idea what really can be extracted out.